src/abstractdataframe/reshape.jl

```diff
@@ -185,39 +193,72 @@ long = stack(wide)
 wide0 = unstack(long)
 wide1 = unstack(long, :variable, :value)
 wide2 = unstack(long, :id, :variable, :value)
+wide3 = unstack(long, [:id, :a], :variable, :value)
 ```
 Note that there are some differences between the widened results above.
-
 """
 function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
-    # `rowkey` integer indicating which column to place along rows
-    # `colkey` integer indicating which column to place along column headers
-    # `value` integer indicating which column has values
-    refkeycol = CategoricalArray{Union{eltype(df[rowkey]), Missing}}(df[rowkey])
-    # make sure we report only levels actually present in rowkey column
-    # this is consistent with how the other unstack method based on groupby works
+    refkeycol = copy(CategoricalArray{Union{eltype(df[rowkey]), Missing}}(df[rowkey]))
     droplevels!(refkeycol)
-    valuecol = df[value]
     keycol = CategoricalArray{Union{eltype(df[colkey]), Missing}}(df[colkey])
+    valuecol = df[value]
+    Nrow = length(refkeycol.pool)
+    Ncol = length(keycol.pool)
+    df2m = [similar_missing(valuecol, Nrow) for i in 1:Ncol]
+    _unstack(df, rowkey, colkey, value, keycol, valuecol, df2m, refkeycol)
+end
+
+function _unstack(df::AbstractDataFrame, rowkey::Int,
+                  colkey::Int, value::Int, keycol, valuecol, df2m, refkeycol)
     Nrow = length(refkeycol.pool)
     Ncol = length(keycol.pool)
-    payload = DataFrame(Any[similar_missing(valuecol, Nrow) for i in 1:Ncol], map(Symbol, levels(keycol)))
-    nowarning = true
+    hadmissing = false # have we encountered missing in refkeycol
+    mask_filled = falses(Ncol, Nrow+1) # has a given [col,row] entry been filled?
+    warned_missing_1 = false # do we print duplicate entries warning
+    warned_missing_2 = false # do we print missing in keycol
+    keycol_order = Vector{Int}(CategoricalArrays.order(keycol.pool))
+    refkeycol_order = Vector{Int}(CategoricalArrays.order(refkeycol.pool))
     for k in 1:nrow(df)
-        j = Int(CategoricalArrays.order(keycol.pool)[keycol.refs[k]])
-        i = Int(CategoricalArrays.order(refkeycol.pool)[refkeycol.refs[k]])
-        if i > 0 && j > 0
-            if nowarning && !ismissing(payload[j][i])
-                warn("Duplicate entries in unstack.")
-                nowarning = false
+        kref = keycol.refs[k]
+        if kref <= 0 # we have found missing in colkey
+            if !warned_missing_2
+                warn("Missing value in '$(_names(df)[colkey])' variable at row $k. Skipping.")
```

Comment:
I've just checked, the convention used in the `DataFrame` constructor is to print column names without quotes. Maybe clearer with "variable X" (rather than "X variable").

Reply:
fixed

Reply:
although now we print "Missing value in variable variable at row"